README.md
# choroplethrCaCensusTract
<!--
v1.0.0: [![Build Status](https://travis-ci.org/arilamstein/choroplethrCaCensusTract.svg?branch=v1.0.0)](https://travis-ci.org/arilamstein/choroplethrCaCensusTract/branches) 

Master: [![Build Status](https://travis-ci.org/arilamstein/choroplethrCaCensusTract.svg)](https://travis-ci.org/arilamstein/choroplethrCaCensusTract)
-->
Shapefile, metadata and visualization functions for US Census Tracts in California

## Development

I blog about this package's development [here](http://www.arilamstein.com/blog).

## Installation

To install the package type the following:

```
# install.packages("devtools")
library(devtools)
install_github("arilamstein/choroplethrCaCensusTract@v1.1.0")
library(choroplethrCaCensusTract)
```

## Documentation

Besides the built-in documentation, you can learn more about this package on the [blog post](http://www.arilamstein.com/blog/2015/06/11/new-package-choroplethrcacensustract/) I wrote about it, which includes several examples. 


NEWS
choroplethrCaCensusTract 1.2.0
----------------------------------------------------------------
-Function get_ca_tract_acs_data now takes a parameter include_moe which, if true,
will return a column with 90% margins of error.

R/data.R
#' Map of US Census Tracts in California
#' 
#' The map comes from the 2010 Cartographic Boundary Shapefiles - Census Tracts 
#' from the US Census Bureau website.
#' A column named "region" was added which is a duplicate of the column GEOID. 
#' The GEOID is the concatenation of the 5-character county FIPS code and the 
#' 6-character TRACTCE column.
#' A column named "county.fips.numeric" was added which is the numeric version
#' of the county FIPS code of the tract.
#'  
#' @seealso \link{ca.tract.regions}
#' @docType data
#' @name ca.tract.map
#' @usage data(ca.tract.map)
#' @references Taken from \url{https://www.census.gov/geo/maps-data/data/cbf/cbf_tracts.html}.
NULL

#' Metadata about the California Census Tracts map
#' 
#' In addition to the region/GEOID it contains the TRACTCE and county FIPS code
#' as both an integer and character (i.e. with and without the leading "0")
#' 
#' @seealso \link{ca.tract.map}
#'  
#' @docType data
#' @name ca.tract.regions
#' @usage data(ca.tract.regions)
NULL

#'  A data.frame containing population estimates for California Census Tracts in 2012. 
#' 
#' @references Taken from the US American Community Survey (ACS) 5 year estimates.
#' @seealso \link{df_ca_tract_demographics}
#'  
#' @docType data
#' @name df_pop_ca_tract
#' @usage data(df_pop_ca_tract)
#' @examples
#' data(df_pop_ca_tract)
#'
#' # zoom in on San Francisco, which has FIPS code 6075
#' ca_tract_choropleth(df_pop_ca_tract, 
#'                     title  = "2012 San Francisco Census Tract\n Population Estimates", 
#'                     legend = "Population",
#'                     county_zoom = 6075)
#' \dontrun{
#' ca_tract_choropleth(df_pop_ca_tract, 
#'                     title  = "2012 California Census Tract\n Population Estimates", 
#'                     legend = "Population")
#' 
#' }
NULL

#' A data.frame containing demographic statistics for each Census Tract in California.
#'       
#' @name df_ca_tract_demographics
#' @docType data
#' @references Data comes from the 2013 5-year American Community Survey (ACS). Data generated by
#' ?get_ca_tract_demographics. 
#' @keywords data
#' @usage data(df_ca_tract_demographics)
#' @examples
#'  # 2013 per capita income estimate
#'  data(df_ca_tract_demographics)
#'  colnames(df_ca_tract_demographics)
#'  
#'  df_ca_tract_demographics$value = df_ca_tract_demographics$per_capita
#'  ca_tract_choropleth(df_ca_tract_demographics,
#'                      title       = "2013 San Francisco Census Tract\n Per Capita Income",
#'                      legend      = "Income",
#'                      num_colors  = 1,
#'                      county_zoom = 6075)
NULL

R/ca_tract_choropleth.R
#' An R6 object for creating choropleths of California Census Tracts.
#' @export
#' @importFrom dplyr left_join
#' @importFrom R6 R6Class
#' @importFrom choroplethr Choropleth
CaTractChoropleth = R6Class("CaTractChoropleth",
  inherit = choroplethr::Choropleth,
  public = list(
    
    # initialize with a map of California Census Tracts
    initialize = function(user.df)
    {
      data(ca.tract.map, package="choroplethrCaCensusTract", envir=environment())
      super$initialize(ca.tract.map, user.df)
      
      if (private$has_invalid_regions)
      {
        warning("Please see ?ca.tract.regions for a list of mappable regions")
      }
    },
    
    # All zooms, at the end of the day, are zip zooms. But often times it is more natural
    # for users to specify the zoom in other geographical units
    # This function name is a bit of a hack - it seems like I cannot override the parent set_zoom directly
    # because this function has a different number of parameters than that function, and the extra parameters
    # seeming just disappear
    set_zoom_tract = function(county_zoom, tract_zoom)
    {
      # user can zoom by at most one of these options
      num_zooms_selected = sum(!is.null(c(county_zoom, tract_zoom)))
      if (num_zooms_selected > 1) {
        stop("You can only zoom in by one of county_zoom or tract_zoom")
      }
      
      data(ca.tract.regions, package="choroplethrCaCensusTract", envir=environment())
      
      # if the zip_zoom field is selected, just do default behavior
      if (!is.null(tract_zoom)) {
        super$set_zoom(tract_zoom)
      # if county_zoom field is selected, extract zips from counties  
      } else if (!is.null(county_zoom)) {
        stopifnot(all(county_zoom %in% unique(ca.tract.regions$county.fips.numeric)))
        tracts = ca.tract.regions[ca.tract.regions$county.fips.numeric %in% county_zoom, "region"]
        super$set_zoom(tracts)        
      }
    }
    
  )
)

#' Create a choropleth of US Census Tracts in California
#' 
#' @param df A data.frame with a column named "region" and a column named "value".  Elements in 
#' the "region" column must exactly match how census tracts are labelled in in the "region" column in ?ca.tract.regions
#' @param title An optional title for the map.  
#' @param legend An optional name for the legend.  
#' @param num_colors The number of colors on the map. A value of 1 
#' will use a continuous scale. A value in [2, 9] will use that many colors. 
#' @param tract_zoom An optional vector of tracts to zoom in on. Elements of this vector must exactly 
#' match the names of tracts as they appear in the "region" column of ?ca.tract.regions.
#' @param county_zoom An optional vector of county FIPS codes to zoom in on. Elements of this 
#' vector must exactly match the names of counties as they appear in the "county.fips.numeric" column 
#' of ?ca.tract.regions.
#' @param reference_map If true, render the choropleth over a reference map from Google Maps.
#'
#' @seealso \url{https://www.census.gov/geo/reference/gtc/gtc_ct.html} for more information on Census Tracts
#' @export
#' @importFrom Hmisc cut2
#' @importFrom stringr str_extract_all
#' @importFrom ggplot2 ggplot aes geom_polygon scale_fill_brewer ggtitle theme theme_grey element_blank geom_text
#' @importFrom ggplot2 scale_fill_continuous scale_colour_brewer  
#' @importFrom scales comma
#' @examples
#' # zoom in on Los Angeles, which has FIPS code 6037
#' data(df_pop_ca_tract)
#' ca_tract_choropleth(df_pop_ca_tract,
#'                     title  = "2012 Los Angeles Census Tract\n Population Estimates",
#'                     legend = "Population",
#'                     county_zoom = 6037)                  
#'
#' # add a reference map
#' ca_tract_choropleth(df_pop_ca_tract,
#'                     title  = "2012 Los Angeles Census Tract\n Population Estimates",
#'                     legend        = "Population",
#'                     county_zoom   = 6037,
#'                     reference_map = TRUE)                  
#'
#' \dontrun{
#' 
#' ca_tract_choropleth(df_pop_ca_tract,
#'                     title  = "2012 California Census Tract\n Population Estimates",
#'                     legend = "Population")
#'
#' # 2013 per capita income estimate
#' data(df_ca_tract_demographics)
#' df_ca_tract_demographics$value = df_ca_tract_demographics$per_capita
#' ca_tract_choropleth(df_ca_tract_demographics,
#'                     title         = "2013 Los Angeles Census Tract\n Per Capita Income",
#'                     legend        = "Income",
#'                     num_colors    = 4,
#'                     county_zoom   = 6037,
#'                     reference_map = TRUE)
#' }
ca_tract_choropleth = function(df, 
                               title         = "", 
                               legend        = "", 
                               num_colors    = 7, 
                               tract_zoom    = NULL, 
                               county_zoom   = NULL, 
                               reference_map = FALSE)
{
  c = CaTractChoropleth$new(df)
  c$title  = title
  c$legend = legend
  c$set_zoom_tract(tract_zoom = tract_zoom, county_zoom = county_zoom)
  c$set_num_colors(num_colors)
  if (reference_map) {
    c$render_with_reference_map()
  } else {
    c$render()
  }
}

R/get_ca_tract_demographics.R
#' Get a handful of demographic variables on California Census Tracts from the US Census Bureau as a data.frame.
#' 
#' The data comes from the American Community Survey (ACS). The variables are: total population, percent White 
#' not Hispanic, Percent Black or African American not Hispanic, percent Asian not Hispanic,
#' percent Hispanic all races, per-capita income, median rent and median age.
#' @param endyear The end year for the survey
#' @param span The span of the survey
#' @references The choroplethr guide to Census data: http://cran.r-project.org/web/packages/choroplethr/vignettes/e-mapping-us-census-data.html
#' @importFrom acs geo.make acs.fetch geography estimate
#' @export
#' @examples
#' \dontrun{
#' df = get_ca_tract_demographics(endyear=2010, span=5)
#' colnames(df)
#'
#' # analyze the percent of people who are white not hispanic
#' # a boxplot shows the distribution
#' boxplot(df$percent_white)
#' 
#' # a choropleth map shows the location of the values in california
#' # set the 'value' column to be the column we want to render
#' df$value = df$percent_white
#' ca_tract_choropleth(df, 
#'                     title="2010 Census Tracts\nPercent White not Hispanic", 
#'                     legend="Percent")
#'
#' # zoom into san francisco county
#' ca_tract_choropleth(df, 
#'                     title="2010 Census Tracts\nPercent White not Hispanic", 
#'                     legend="Percent",
#'                     county_zoom=6075)
#' }
get_ca_tract_demographics = function(endyear=2013, span=5)
{  
  all.ca.tracts = get_all_ca_tracts()
  race.data = acs::acs.fetch(geography    = all.ca.tracts, 
                             table.number = "B03002", 
                             col.names    = "pretty", 
                             endyear      = endyear, 
                             span         = span)
  
  # dummy to get proper regions
  dummy.df = convert_acs_obj_to_df(race.data, 1)
  
  # convert to a data.frame 
  df_race = data.frame(region                   = dummy.df$region,  
                       total_population         = as.numeric(acs::estimate(race.data[,1])),
                       white_alone_not_hispanic = as.numeric(acs::estimate(race.data[,3])),
                       black_alone_not_hispanic = as.numeric(acs::estimate(race.data[,4])),
                       asian_alone_not_hispanic = as.numeric(acs::estimate(race.data[,6])),
                       hispanic_all_races       = as.numeric(acs::estimate(race.data[,12])))
  
  df_race$region = as.character(df_race$region) # no idea why, but it's a factor before this line
  
  df_race$percent_white    = round(df_race$white_alone_not_hispanic / df_race$total_population * 100)
  df_race$percent_black    = round(df_race$black_alone_not_hispanic / df_race$total_population * 100)
  df_race$percent_asian    = round(df_race$asian_alone_not_hispanic / df_race$total_population * 100)
  df_race$percent_hispanic = round(df_race$hispanic_all_races       / df_race$total_population * 100)
  
  df_race = df_race[, c("region", "total_population", "percent_white", "percent_black", "percent_asian", "percent_hispanic")]
  
  # per capita income 
  df_income = get_ca_tract_acs_data("B19301", endyear=endyear, span=span)[[1]]   
  colnames(df_income)[[2]] = "per_capita_income"
  
  # median rent
  df_rent = get_ca_tract_acs_data("B25058", endyear=endyear, span=span)[[1]]  
  colnames(df_rent)[[2]] = "median_rent"
  
  # median age
  df_age = get_ca_tract_acs_data("B01002", endyear=endyear, span=span, column_idx=1)[[1]]  
  colnames(df_age)[[2]] = "median_age"
  
  df_demographics = merge(df_race        , df_income, all.x=TRUE)
  df_demographics = merge(df_demographics, df_rent  , all.x=TRUE)  
  df_demographics = merge(df_demographics, df_age   , all.x=TRUE)
    
  df_demographics
}

R/acs.R
#' @importFrom acs geo.make acs.fetch geography
get_all_ca_tracts = function()
{
  ca.counties = acs.fetch(geography    = geo.make(state = 6, county = "*"), 
                          table.number = "B01003")
  
  geo.make(state  = "CA", 
           county = as.numeric(geography(ca.counties)[[3]]), 
           tract  = "*")
}

#' Returns a list representing American Community Survey (ACS) estimates
#'
#' Given an ACS tableId, endyear and span. Prompts user for the column id if there 
#' are multiple tables. The first element of the list is a data.frame with estimates. 
#' The second element is the ACS title of the column.
#' Requires the acs package to be installed, and a Census API Key to be set with the 
#' acs's api.key.install function.  Census API keys can be obtained at http://api.census.gov/data/key_signup.html.
#'
#' @param tableId The id of an ACS table
#' @param endyear The end year of the survey to use.  See acs.fetch (?acs.fetch) and http://1.usa.gov/1geFSSj for details.
#' @param span The span of time to use.  See acs.fetch and http://1.usa.gov/1geFSSj for details.
#' on the same longitude and latitude map to scale. This variable is only checked when the "states" variable is equal to all 50 states.
#' @param column_idx The optional column id of the table to use. If not specified and the table has multiple columns,
#' you will be prompted for a column id.
#' @param include_moe Whether to include the 90 percent margin of error. 
#' @export
#' @seealso http://factfinder2.census.gov/faces/help/jsf/pages/metadata.xhtml?lang=en&type=survey&id=survey.en.ACS_ACS, which lists all ACS Surveys.
#' @importFrom acs acs.fetch geography estimate geo.make
get_ca_tract_acs_data = function(tableId, endyear=2012, span=5, column_idx=-1, include_moe=FALSE)
{
  acs.data = acs.fetch(geography    = get_all_ca_tracts(), 
                       table.number = tableId,
                       col.names    = "pretty", 
                       endyear      = endyear, 
                       span         = span)

  if (column_idx == -1) {
    column_idx = get_column_idx(acs.data, tableId) # some tables have multiple columns 
  }
  title      = acs.data@acs.colnames[column_idx] 
  df         = convert_acs_obj_to_df(acs.data, column_idx, include_moe) # choroplethr requires a df
  list(df=df, title=title) # need to return 2 values here
}
  
# support multiple column tables
get_column_idx = function(acs.data, tableId)
{
  column_idx = 1
  if (length(acs.data@acs.colnames) > 1)
  {
    num_cols   = length(acs.data@acs.colnames)
    title      = paste0("Table ", tableId, " has ", num_cols, " columns.  Please choose which column to render:")
    column_idx = menu(acs.data@acs.colnames, title=title)
  }
  column_idx
}

# the acs package returns data as a custom S4 object. But we need the data as a data.frame.
# with 1 column named "region". The map requires the region to be 11 character - 
# 5 for the county FIPS code and 6 for the tract
#' @importFrom acs standard.error
convert_acs_obj_to_df = function(acs.data, column_idx, include_moe) 
{
  df = data.frame(state  = geography(acs.data)$state,   # integer
                  county = geography(acs.data)$county,  # integer
                  tract  = geography(acs.data)$tract,   # character
                  value  = as.numeric(estimate(acs.data[,column_idx])))
  
  if (include_moe)
  {
    df$margin.of.error = 1.645 * as.numeric(standard.error(acs.data[, column_idx])) 
  }    
  
  # county fips code must be 5 chars
  # 2 chars for the state (i.e. leading "0")
  df$state = as.character(df$state)
  df$state = paste0("0", df$state)
  # 3 chars for the county - i.e. leading "0" or leading "00"
  df$county = as.character(df$county)
  for (i in 1:nrow(df))
  {
    if (nchar(df[i, "county"]) == 1) {
      df[i, "county"] = paste0("00", df[i, "county"])
    } else if (nchar(df[i, "county"]) == 2) {
      df[i, "county"] = paste0("0", df[i, "county"])
    }
  } 

  # now concat with the tract id
  df$region = paste0(df$state, df$county, df$tract)

  # only include relevant columns
  if (include_moe)
  {
    df[, c("region", "value", "margin.of.error")] # only return (region, value) pairs
  } else {
    df[, c("region", "value")]
  }
}

#' Create a choropleth map of California Census Tracts from ACS data
#' 
#' Creates a choropleth of California Census Tracts the US Census' American Community Survey (ACS) data.  
#' Requires the acs package to be installed, and a Census API Key to be set with 
#' the acs's api.key.install function.  Census API keys can be obtained at http://www.census.gov/developers/tos/key_request.html.
#'
#' @param tableId The id of an ACS table
#' @param endyear The end year of the survey to use.  See acs.fetch (?acs.fetch) and http://1.usa.gov/1geFSSj for details.
#' @param span The span of time to use.  See acs.fetch and http://1.usa.gov/1geFSSj for details.
#' @param num_colors The number of colors on the map. A value of 1 
#' will use a continuous scale. A value in [2, 9] will use that many colors. 
#' @param tract_zoom An optional list of states to zoom in on. Must come from the "region" column in
#' ?ca.tract.regions.
#' @param county_zoom An optional list of counties to zoom in on. Must match entries of the county.fips.numeric column of 
#' ?ca.tract.regions.
#' @return A choropleth.
#' 
#' @keywords choropleth, acs
#' 
#' @seealso \code{api.key.install} in the acs package which sets an Census API key for the acs library
#' @references The choroplethr guide to Census data: http://cran.r-project.org/web/packages/choroplethr/vignettes/e-mapping-us-census-data.html
#' @references Uses the acs package created by Ezra Haber Glenn.
#' @export
#' @importFrom acs acs.fetch geography estimate geo.make
#' @examples
#' \dontrun{
#' # per capita income, san francisco (fips code 6075)
#' ca_tract_choropleth_acs("B19301", county_zoom=6075)
#' 
#' # aggregate public assistance income, san francisco
#' ca_tract_choropleth_acs("B19067", county_zoom=6075)
#' }
ca_tract_choropleth_acs = function(tableId, endyear=2011, span=5, num_colors=7, tract_zoom=NULL, county_zoom=NULL)
{
  acs.data = get_ca_tract_acs_data(tableId, endyear, span)
  ca_tract_choropleth(acs.data[['df']], acs.data[['title']], "", num_colors, tract_zoom, county_zoom)
}